design-proposals/klip-24-key-column-semantics-in-queries.md

```diff
@@ -0,0 +1,574 @@
+# KLIP 24 - KEY column semantics in queries
+
+**Author**: @big-andy-coates |
+**Release Target**: TBD |
+**Status**: In Discussion |
+**Discussion**: TBD
+
+**tl;dr:** Persistent queries do not allow the key column in the projection as key columns are
+currently _implicitly_ copied across. This is not intuitive to anyone familiar with SQL. This KLIP
+proposes flipping this so that the key column is _not_ implicitly copied across.
+
+## Motivation and background
+
+Below, we contrast the differences between transient and persistent queries and their current key
+column handling semantics.
+
+Note: in all the example sql in this KLIP:
+ * `EMIT CHANGES` had been removed for brevity.
+ * `ROWTIME` had been ignored for brevity.
+ * Schemas use `KEY` to denote the column stored in the Kafka record's key.
+ * `PRIMARY KEY` is not used for brevity.
+ * All input sources use the schema `ID INT KEY, V0 INT, V1 INT`, unless otherwise stated.
+
+Transient push and pull queries only return the columns in their projection. This is inline with
+standard SQL. Conversely, persistent queries _implicitly_ copy the key column across, and don't
+allow the key column in the projection.
+
+### Simple query
+
+The difference in key column handling can be seen with a simple query:
+
+```sql
+SELECT ID, V0, V1 FROM INPUT;
+-- resulting columns: ID INT, V0 INT, V1 INT
+
+-- vs --
+
+CREATE TABLE OUTPUT AS
+   SELECT ID, V0, V1 FROM INPUT;
+-- fails with error about duplicate 'ID' column.
+```
+
+Where as the above transient query works and returns all columns, converting it to a persistent
+query causes it to fail with a duplicate column error, because `ID` is _implicitly_ being copied to
+the OUTPUT schema, and the `ID` in the projection is creating a value column called `ID`. That's two
+columns named `ID` and hence the query fails.
+
+We propose that the above `CREATE TABLE` statement should work, as the projection is equivalent to
+`select *`, which works.
+
+### Non-join select star
+
+For non-join queries, both transient and persistent queries select all the columns in the schema:
+
+```sql
+SELECT * FROM INPUT;
+-- resulting columns: ID INT, V0 INT, V1 INT
+
+-- vs --
+
+CREATE TABLE OUTPUT AS
+   SELECT * FROM INPUT;
+-- resulting schema: ID INT KEY, V0 INT, V1 INT (Same as above)
+```
+
+### Joins
+
+The differences with joins is more subtle.
+
+With a join query, a transient query selects all columns from all sources. A persistent query adds
+all columns from all sources, and also adds an additional column that stores the result of the join
+criteria.
+
+```sql
+SELECT * FROM INPUT I1 JOIN INPUT_2 I2 ON I1.ID = I2.ID;
+-- resulting columns: I1_ID INT, I1_V0 INT, I1_V1 INT, I2_ID INT, I2_V0 INT, I2_V1 INT
+
+-- vs --
+
+CREATE TABLE OUTPUT AS
+   SELECT * FROM INPUT I1 JOIN INPUT_2 I2 ON I1.ID = I2.ID;
+-- resulting schema (any key name enabled): ID INT KEY, I1_ID INT, I1_V0 INT, I1_V1 INT, I2_ID INT, I2_V0 INT, I2_V1 INT
```

Comment:
Where does the name `ID` come from? It's not specified in the query via `AS ID`?

Reply:
This is existing functionality, not proposed functionality.  The `ID` comes from `I1.ID`, but has the source removed. It can't be called `I1_ID` as that could clash with the column of the value column of the same name.

This KLIP proposes it should be `I1_ID` and the value column should be removed.

Reply:
My point is: the input schemas have 3 columns each, and thus the output schema should have 6 columns, not 7.

Instead of maintaining `I1.ID` twice, ie, in the value as `I1_ID` and in the key as `ID` we just "move" the `I1.ID` column into the key only and don't have it in the value:
```
output schema: I1_ID INT KEY, I1_V0 INT, I1_V1 INT, I2_ID INT, I2_V0 INT, I2_V1 INT
```

This seems to be much easier to understand?

Reply:
That's exactly what I'm saying the KLIP is proposing. 

You're reading the section on 'current behaviour' and not the section on 'proposed behaviour'!!!

Reply:
I thought that the current behavior would add a column `ROWKEY` (as only this KLIP would allow users to set a different name(?